ipc_receivefile.c

```diff
@@ -2,126 +2,146 @@
+				FILE *fp = fopen(info->file_name, "wb");
+				if (fp == 0) {
+					exit(EXIT_FAILURE);
+				}
 				// Create pipe and wait for writer to be ready
-				mkfifo(method_name, 0777);
+				mkfifo(info->method_name, 0777);
 				printf("Waiting for writer\n");
-				int fd = open(method_name, O_RDONLY);
+				int fd_pipe = open(info->method_name, O_RDONLY);
 
-				if (fd > 0)
+				if (fd_pipe > 0)
 				{
 					printf("Writer found\n");
-					while (num = read(fd, line, sizeof(line)))
+
+					while (num = read(fd_pipe, line, sizeof(line)))
 					{
 						fwrite(line, 1, num, fp);
 					}
 					printf("End of file\n");
-					close(fd);
+					close(fd_pipe);
 				} else {
 					printf("Failed to create pipe");
 					exit(EXIT_FAILURE);
 				}
+				fclose(fp);
+			} else if (info->qflag) {
+				fd = open(info->file_name, O_RDWR | O_CREAT | O_EXCL, 0660);
+				if (fd == -1){
+					perror("open");
+					exit(EXIT_FAILURE);
+				}
+
+				printf("<queue is used, queue name is %s>\n", info->method_name);
+				printf("Receive file name: %s\n", info->file_name);
 				
-			} else if (qflag) {
-				printf("<queue is used, queue name is %s>\n", method_name);
-				printf("queue has not been modified yet!!!\n");
+				mqd_t mq ;
+				int read_size = 0;
+
+				struct mq_attr new_attr;
+				struct timespec	ts;
+									
+				mq = mq_open(info->method_name, O_RDONLY | O_CREAT , 0666, NULL);
+				
+				
+				unsigned int rvprio = MQ_PRIO_MAX-1;
+
+				mq_getattr(mq, &new_attr);
+				long int mq_msgsize = new_attr.mq_msgsize;
+				printf("message size:  %ld\n", new_attr.mq_msgsize);
+
+				char *data = malloc(mq_msgsize);
+				if (data == NULL)
+				{
+					perror("malloc");
+					mq_close(mq);
+					mq_unlink(info->method_name);
+					exit(EXIT_FAILURE);
+				}
+				
+				int total_size = 0;
+				
+				while (read_size != -1)
+				{
+					clock_gettime(CLOCK_REALTIME, &ts);
+					ts.tv_sec += 7; /* set timeout for 7 seconds */
+					ts.tv_nsec = 0; /* Invalid */
+
+					//error_code = errno;
+					read_size = mq_timedreceive(mq, data, mq_msgsize, &rvprio, &ts);
+					if (read_size < 0){
+						if (errno == ETIMEDOUT) /* Timeout error */
+						{
+							// timeout
+							fprintf(stdout, "waiting for receive msg...\n");
+							continue;
+						}
+
+						fprintf(stderr, "mq_timedreceive failed receive size:[%d], errno:[%d]\n", read_size, errno);
+						mq_close(mq);
+						mq_unlink(info->method_name);
+						exit(EXIT_FAILURE);
+						
+					}else {
+						write_to_file(data, read_size, fd);
+						total_size += read_size;
+
+					}
+				}
+
+				if (total_size == find_file_size(fd))
+				{
+					printf("File received | File size = %ld byte(s)\n", find_file_size(fd));
+					free(data);
+					mq_close(mq);
+					mq_unlink(info->method_name);
+					close(fd);
+				} else{
+					printf("File can not received | File size = %d byte(s)\n", total_size);
+					mq_close(mq);
+					mq_unlink(info->method_name);
+					close(fd);
+					exit(EXIT_FAILURE);
+				}	
```

Comment:
Redundancy + missing deallocs.

```suggestion
                free(data);
                mq_close(mq);
				mq_unlink(info->method_name);
				close(fd);
				
				if (total_size == find_file_size(fd))
				{
					printf("File received | File size = %ld byte(s)\n", find_file_size(fd));

				} else{
					printf("File can not received | File size = %d byte(s)\n", total_size);
					free(info);
					exit(EXIT_FAILURE);
				}	
```